### ISSUE TITLE
undefined index notice

### ISSUE THREAD

**Issue by michaeltorbert** [semperfiwebdesign/all-in-one-seo-pack#332]
https://wordpress.org/support/topic/notice-undefined-index-aiosp_can_set_protocol?replies=4

**Comment by michaeltorbert** [comment]
https://wordpress.org/support/topic/noticeunder-fined-index?replies=1

**Comment by michaeltorbert** [comment]
In validate_url_scheme, check for isset... is there a built-in WP function to help with the preg_replace?

**Comment by michaeltorbert** [comment]
For some reason, all the options aren't saving on install (at least in some cases) and MAYBE even after clicking save in some cases.

https://wordpress.org/support/topic/noticeunder-fined-index
https://wordpress.org/support/topic/notice-undefined-index-aiosp_can_set_protocol

**Comment by michaeltorbert** [comment]
Additionally, add back in the isset check.

**Comment by michaeltorbert** [comment]
I just want to people out that the issue isn't an undefined index notice, it's that all the settings weren't being saved with their defaults on a new install.

**Comment by michaeltorbert** [comment]
This has to be tested very carefully. 

I think the underlying issue here may be fixed. For testing, ignore whether you get a notice (although that would be helpful). Instead, look at what settings are set after installing fresh. Make sure all the settings installed now are the ones from before we broke this (maybe around 3.1?)

Once we've confirmed this is fixed, we need to add back in the isset check. (But not until then, because we'll just be covering up the problem.)

**Comment by michaeltorbert** [comment]
Now, we're unable to reproduce the issue on 2.3.5.1. All the settings seem to go in immediately. Maybe it's a multisite problem? At least one of the people reporting the problem had multisite.

**Comment by michaeltorbert** [comment]
Ok so defaults do save on new installs. The issue with one person is that if you're coming from pre-2.0, any new settings may not have saved their defaults if you never configured something and clicked save after that.
For the other people, they're getting not defined because by default that setting is "Auto" in the UI, but "Auto" in the UI doesn't actually set it as anything in the database... separate issues, it just so happened that this issue revealed the other one for the pre-2.0 setup. @wpsmort  We could potentially run an update script that applies the defaults for anyone who doesn't have them... or we could just leave well enough alone. There can't be too many people coming from 1.6.x these days, and if they even one time change a setting and click save on the general settings page (or probably anywhere else) I think they should get all the defaults they're missing (I think, although this is worth testing).

**Comment by michaeltorbert** [comment]
BTW, this is what his settings export looks like. Notice the things that haven't existed in 4 or 5 years, like aiosp_enabled (proof he's been running it a long time). And missing almost anything new except things that are explicitly put in like last active version, active modules...

; Settings export file for All in One SEO Pack

[aioseop_options]

aiosp_can = 'on'
aiosp_donate = 'on'
aiosp_home_title = 'REMOVED FOR PRIVACY'
aiosp_home_description = 'REMOVED FOR PRIVACY'
aiosp_home_keywords = 'REMOVED FOR PRIVACY'
aiosp_max_words_excerpt = NULL
aiosp_rewrite_titles = 'on'
aiosp_post_title_format = '%post_title% | %blog_title%'
aiosp_page_title_format = '%page_title% | %blog_title%'
aiosp_category_title_format = '%category_title% | %blog_title%'
aiosp_tag_title_format = '%tag% | %blog_title%'
aiosp_search_title_format = '%search% | %blog_title%'
aiosp_description_format = '%description%'
aiosp_404_title_format = 'Nothing found for %request_words%'
aiosp_paged_format = ' - Part %page%'
aiosp_use_categories = 'on'
aiosp_dynamic_postspage_keywords = 'on'
aiosp_category_noindex = 'on'
aiosp_tags_noindex = NULL
aiosp_cap_cats = 'on'
aiosp_generate_descriptions = 'on'
aiosp_debug_info = NULL
aiosp_post_meta_tags = ''
aiosp_page_meta_tags = ''
aiosp_home_meta_tags = ''
aiosp_enabled = '1'
aiosp_do_log = NULL
aiosp_ex_pages = ''
aiosp_use_tags_as_keywords = 'on'
aiosp_posttypecolumns = 'a:2:{i:0;s:4:"post";i:1;s:4:"page";}'
aiosp_enablecpost = NULL
aiosp_seopostcol = NULL
aiosp_seocustptcol = NULL
aiosp_archive_author_noindex = 'on'
aiosp_archive_date_noindex = 'on'
aiosp_token = '4/jAhekSMgT07Njj0oUwXScvY3pIYrr5s3nEdxWvwte4s'
aiosp_secret = 'B83XR9yzL7tbyBIb6A_iHfS5'
aiosp_date_title_format = NULL
last_active_version = '2.3.5.1'

[aiosp_feature_manager_options]

aiosp_feature_manager_enable_sitemap = ''
aiosp_feature_manager_enable_opengraph = ''
aiosp_feature_manager_enable_robots = ''
aiosp_feature_manager_enable_file_editor = ''
aiosp_feature_manager_enable_importer_exporter = 'on'
aiosp_feature_manager_enable_bad_robots = ''
aiosp_feature_manager_enable_performance = 'on'

**Comment by wpsmort** [comment]
The original undefined index notice issue has been fixed and all future work regarding applying defaults will be done under issue #364
